### PR TITLE
upgrade to 0.6+

### PR DESCRIPTION
**Pull request by jrynyt**
Work with Drone 0.6+, while still working with 0.4, but not 0.5 (because it's deprecated, according to the maintainers, and secrets are totally different than the other two).

Tested with `action: update` on:

0.4: https://drone-general-dev.newsdev.net/nytm/jryan-test/15

0.7: https://drone.dv.nyt.net/nytm/jryan-test/17

and added a few unit tests.

I'm not really familiar with all the ways people are using this, so other testing would be appreciated. This is available now on `jrynyt2/drone-gae:latest`

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -91,55 +91,35 @@ func wrapMain() error {
 
 	fmt.Printf("Drone GAE Plugin built from %s\n", rev)
 
-	// https://godoc.org/github.com/drone/drone-plugin-go/plugin
-	workspace := plugin.Workspace{}
-	repo := plugin.Repo{}
-	build := plugin.Build{}
-	system := plugin.System{}
```
Comment (jrynyt) [inline main.go]: none of these 3 were actually used, so they're not being copied into configFromStdin()

File: main.go [hunk 2]
```diff
@@ -154,7 +134,7 @@ func wrapMain() error {
 		}
 	}()
 
-	runner := NewEnviron(filepath.Join(workspace.Path, vargs.Dir), os.Environ(),
+	runner := NewEnviron(filepath.Join(workspace, vargs.Dir), os.Environ(),
```
Comment (jrynyt) [inline main.go]: the only property of `plugin.Workspace{}` that was being used was `Path`, so `workspace` is just becoming a string to fit the new built-in drone variables.

File: main_test.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		Action:  "dostuff",
+	}
+	assert.NoError(t, validateVargs(&vargs))
+	// Project field overrides token
+	assert.Equal(t, "my-other-project", vargs.Project)
+}
+
+/*
+//This works in regular go, but not under go test for some reason
+//plugin.MustParse() panics due to stdin EOF
```
Comment (jrynyt) [inline main_test.go]: I'm open to suggestions to making this work

File: main.go [hunk 4]
```diff
@@ -91,55 +91,35 @@ func wrapMain() error {
...
-	if vargs.Project == "" {
-		return fmt.Errorf("missing required param: project")
+	// Check what drone version we're running on
+	if os.Getenv("DRONE_WORKSPACE") == "" { // 0.4
+		err := configFromStdin(&vargs, &workspace)
+		if err != nil {
+			return err
+		}
+	} else { // 0.5+
+		err := configFromEnv(&vargs, &workspace)
```
Comment (jrynyt) [inline main.go]: this paradigm of passing pointers to things that the function writes into (vs return values) is new to me, so let me know if this isn't right.

File: exec.go [hunk 5]
```diff
@@ -23,6 +28,8 @@ func NewEnviron(dir string, env []string, stdout, stderr io.Writer) *Environ {
 
 // Run executes the given program.
 func (e *Environ) Run(name string, arg ...string) error {
+	displayArg := reRedact.ReplaceAll([]byte(strings.Trim(fmt.Sprint(arg), "[]")), []byte("$1 [redacted] "))
```
Comment (jprobinson) [inline exec.go]: Looks like there's an unneeded `fmt.Sprint` on the arg here.

Comment (jprobinson) [reply exec.go]: nvm, you're just grabbing the stringer of the array.

Carry on!

File: main.go [hunk 6]
```diff
@@ -173,14 +153,105 @@ func wrapMain() error {
 
 }
 
+func configFromStdin(vargs *GAE, workspace *string) error {
+	// https://godoc.org/github.com/drone/drone-plugin-go/plugin
+	workspaceInfo := plugin.Workspace{}
+	plugin.Param("workspace", &workspaceInfo)
+	plugin.Param("vargs", vargs)
+	// Note this hangs if no cli args or input on STDIN
+	plugin.MustParse()
```
Comment (jrynyt) [inline main.go]: this just [panics if something gets messed up](https://github.com/drone/drone-plugin-go/blob/master/plugin/param.go#L126-L131), so no error to handle nicely

File: main.go [hunk 7]
```diff
@@ -173,14 +153,105 @@ func wrapMain() error {
...
+	*workspace = os.Getenv("DRONE_WORKSPACE")
+	vargs.Token = os.Getenv("GAE_CREDENTIALS") // secrets are not prefixed
+	vargs.Version = os.Getenv("PLUGIN_VERSION")
+	vargs.FlexImage = os.Getenv("PLUGIN_FLEX_IMAGE")
+	vargs.AppFile = os.Getenv("PLUGIN_APP_FILE")
+	vargs.CronFile = os.Getenv("PLUGIN_CRON_FILE")
+	vargs.DispatchFile = os.Getenv("PLUGIN_DISPATCH_FILE")
+	vargs.Dir = os.Getenv("PLUGIN_DIR")
+	vargs.AppCfgCmd = os.Getenv("PLUGIN_APPCFG_CMD")
+	vargs.GCloudCmd = os.Getenv("PLUGIN_GCLOUD_CMD")
```
Comment (jprobinson) [inline main.go]: Next time, you may save some finger work with this lib: https://github.com/kelseyhightower/envconfig

**Comment by jrynyt** [comment]
Thanks for the quick review @jprobinson -- could you test with your setup whenever you get a chance?

**Comment by jprobinson** [comment]
Not sure how I'd be able to test until we have this in master/dockerhub, right?

**Comment by jrynyt** [comment]
this version is in dockerhub now, and can be used with `image: jrynyt2/drone-gae:latest` in .drone.yml

**Comment by jrynyt** [comment]
thanks @fsouza for testing this too: https://drone.dv.nyt.net/nytm/messaging-helix-webhooks/3

**Pull request merged**